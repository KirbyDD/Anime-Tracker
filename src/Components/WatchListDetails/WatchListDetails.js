import React, { Component } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import {
  updateCurrentEpisode,
  updateCurrentComment,
  removeShow
} from "../../actions/index";
import "./WatchListDetails.scss";
import PropTypes from "prop-types";

export class WatchListDetails extends Component {
  constructor(props) {
    super(props);
    this.state = {
      epBtn: true,
      form: false,
      commentBtn: true,
      comment: false
    };
  }

  numOptions = () => {
    let options = [];
    for (let i = 1; i <= this.props.episodes; i++) {
      options.push(<option value={i}>{i}</option>);
      if (options.length === this.props.episodes) {
        return options;
      }
    }
  };

  titleToDisplay = () => {
    if (this.props.enTitle) {
      return this.props.enTitle;
    } else {
      return this.props.jpTitle;
    }
  };

  changeCurrentEpisode = () => {
    this.setState({
      epBtn: false,
      form: true
    });
  };

  changeComment = () => {
    this.setState({
      commentBtn: false,
      comment: true
    });
  };

  updateEpisode = (id, value) => {
    this.props.updateEpisode(id, value);
    this.setState({
      epBtn: true,
      form: false
    });
  };

  updateComment = (id, value) => {
    this.props.updateComment(id, value);
    this.setState({
      commentBtn: true,
      comment: false
    });
  };

  setCurrentEpisode = () => {
    let show = this.props.followedShows.find(show => show.id === this.props.id);
    return show.currentEpisode;
  };

  setComment = () => {
    let show = this.props.followedShows.find(show => show.id === this.props.id);
    return show.comment;
  };

  render() {
    let title = this.titleToDisplay();
    let currentEpisode = this.setCurrentEpisode();
    let comment = this.setComment();
    return (
      <section className="watchlist-details">
        <div className="watchlist-detail-info">
          <h2>{title}</h2>
          <h3>{this.props.synopsis}</h3>
          <p className="detail-text">What Episode are you currently on?</p>
          <p className="detail-text">{currentEpisode}</p>
          {this.state.epBtn && (
            <button id='epBtn' onClick={e => this.changeCurrentEpisode()}>Update</button>
          )}
          {this.state.form && (
            <select
              id='update-epBtn'
              onChange={e => this.updateEpisode(this.props.id, e.target.value)}
              selected={this.props.currentEpisode}
            >
              {this.numOptions()}
            </select>
          )}
          <p className="detail-text">Reminder Section</p>
          <p className="detail-text">{comment}</p>
          {this.state.commentBtn && (
            <button id='commentBtn' onClick={e => this.changeComment()}>Update</button>
          )}
          {this.state.comment && (
            <article>
              <textarea placeholder="Type new reminder here" />
              <br />
              <button
                id='update-commentBtn'
                onClick={e =>
                  this.updateComment(
                    this.props.id,
                    e.target.previousElementSibling.previousElementSibling.value
                  )
                }
              >
                Sumbit
              </button>
            </article>
          )}
          <br />
          <Link to="/watchlist">
            <button
              className="remove-btn"
              onClick={e => this.props.removeFromWatchList(this.props.id)}
            >
              Remove From WatchList
            </button>
          </Link>
        </div>
        <img className="side-image" src={this.props.posterImage.large} />
      </section>
    );
  }
}

export const mapStateToProps = store => ({
  ...store
});

export const mapDispatchToProps = dispatch => ({
  updateEpisode: (id, value) => dispatch(updateCurrentEpisode(id, value)),
  updateComment: (id, comment) => dispatch(updateCurrentComment(id, comment)),
  removeFromWatchList: id => dispatch(removeShow(id))
});

WatchListDetails.propTypes = {
  anime: PropTypes.array,
  comment: PropTypes.string,
  coverImage: PropTypes.object,
  currentEpisode: PropTypes.number,
  enTitle: PropTypes.string,
  episodes: PropTypes.number,
  followedShows: PropTypes.array,
  id: PropTypes.number,
  jpTitle: PropTypes.string,
  posterImage: PropTypes.object,
  removeFromWatchList: PropTypes.func,
  synopsis: PropTypes.string,
  updateComment: PropTypes.func,
  updateEpisode: PropTypes.func
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(WatchListDetails);
