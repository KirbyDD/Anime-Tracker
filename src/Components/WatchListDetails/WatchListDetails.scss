.watchlist-details {
  display: flex;
}

.watchlist-detail-info {
  padding: 30px;
  text-align: center;
}

.detail-text {
  font-size: 25px;
}

.remove-btn {
  margin-top: 35px;
  width: 250px;
}

.side-image {
  height: 100%;
}

textarea {
  font-size: 1.4rem;
  height: 100px;
  width: 500px;
}
