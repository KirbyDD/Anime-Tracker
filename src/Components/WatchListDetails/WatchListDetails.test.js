import React from 'react';
import {shallow} from 'enzyme';
import {WatchListDetails} from './WatchListDetails';
import { exportAllDeclaration } from '@babel/types';

let props = {
  id: 1,
  synopsis: 'This is the synopsis',
  enTitle: 'Life of TronKat',
  jpTitle: 'Tronkat no Life desu',
  episodes: 27,
  posterImage: {
    large: 'https://i.imgur.com/vihivug.jpg',
    tiny: null
  },
  coverImage: {
    small: 'https://i.imgur.com/vihivug.jpg',
    tiny: null
  },
  currentEpisode: 1,
  comment: ''
}

let props2 = {
  id: 1,
  synopsis: 'This is the synopsis',
  enTitle: null,
  jpTitle: 'Tronkat no Life desu',
  episodes: 27,
  posterImage: {
    large: 'https://i.imgur.com/vihivug.jpg',
    tiny: null
  },
  coverImage: {
    small: 'https://i.imgur.com/vihivug.jpg',
    tiny: null
  },
  currentEpisode: 7,
  comment: 'Testing'
}

let followedShows = [
  {
    id: 1,
    synopsis: 'This is the synopsis',
    enTitle: 'Life of TronKat',
    jpTitle: 'Tronkat no Life desu',
    episodes: 27,
    posterImage: {
      large: 'https://i.imgur.com/vihivug.jpg',
      tiny: null
    },
    coverImage: {
      small: 'https://i.imgur.com/vihivug.jpg',
      tiny: null
    },
    rating: '99.9',
    startDate: '1990-01-26',
    endDate: 'Still Playing',
    ageRating: '17+'
  }
]

let followedShows2 = [
  {
    id: 1,
  synopsis: 'This is the synopsis',
  enTitle: null,
  jpTitle: 'Tronkat no Life desu',
  episodes: 27,
  posterImage: {
    large: 'https://i.imgur.com/vihivug.jpg',
    tiny: null
  },
  coverImage: {
    small: 'https://i.imgur.com/vihivug.jpg',
    tiny: null
  },
  currentEpisode: 7,
  comment: 'Testing'
  }
]

describe('WatchList Details', () => {
  let wrapper;
  let updateEpisode;
  let updateComment;

  beforeEach(function() {
    updateEpisode = jest.fn();
    updateComment = jest.fn();
    wrapper = shallow(
      <WatchListDetails {...props} episodes={20}
                        followedShows={followedShows}
                        updateEpisode={updateEpisode}
                        updateComment={updateComment}/>
    )
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should match snapshot', () => {
    wrapper.instance().changeCurrentEpisode();
    expect(wrapper).toMatchSnapshot();
  })

  it('should match snapshot', () => {
    wrapper.instance().changeComment();
    expect(wrapper).toMatchSnapshot();
  })

  it('should have default state properties', () => {
    expect(wrapper.state()).toEqual({
      epBtn: true,
      form: false,
      commentBtn: true,
      comment: false
    })
  })

  it('should have update state when Changing Episode', () => {
    wrapper.instance().changeCurrentEpisode();
    expect(wrapper.state()).toEqual({
      epBtn: false,
      form: true,
      commentBtn: true,
      comment: false
    })
  })

  it('should have update state when Updating episode', () => {
    wrapper.instance().changeCurrentEpisode();
    expect(wrapper.state()).toEqual({
      epBtn: false,
      form: true,
      commentBtn: true,
      comment: false
    })

    wrapper.instance().updateEpisode();
    expect(wrapper.state()).toEqual({
      epBtn: true,
      form: false,
      commentBtn: true,
      comment: false
    })
  })

  it('should have update state when Changing Comment', () => {
    wrapper.instance().changeComment();
    expect(wrapper.state()).toEqual({
      epBtn: true,
      form: false,
      commentBtn: false,
      comment: true
    })
  })

  it('should have update state Updating Comment', () => {
    wrapper.instance().changeComment();
    expect(wrapper.state()).toEqual({
      epBtn: true,
      form: false,
      commentBtn: false,
      comment: true
    })
    wrapper.instance().updateComment()
    expect(wrapper.state()).toEqual({
      epBtn: true,
      form: false,
      commentBtn: true,
      comment: false
    })
  })

  it('should return English Title', () => {
    expect(wrapper.instance().titleToDisplay()).toEqual('Life of TronKat')
  })

  it('should return japanese Title', () => {
    wrapper = shallow(
      <WatchListDetails {...props2} episodes={20}
      followedShows={followedShows}
      updateEpisode={updateEpisode}
      updateComment={updateComment}/>
      )
    expect(wrapper.instance().titleToDisplay()).toEqual('Tronkat no Life desu')
  })

  it('should return an array whose length is equal to episodes', () => {
    expect(wrapper.instance().numOptions().length).toEqual(20)
  })

  it('should return current episode', () => {
    wrapper = shallow(
      <WatchListDetails {...props2} episodes={20}
      followedShows={followedShows2}
      updateEpisode={updateEpisode}
      updateComment={updateComment}/>
      )
    expect(wrapper.instance().setCurrentEpisode()).toEqual(7)
  })

  it('should return current reminder comment', () => {
    wrapper = shallow(
      <WatchListDetails {...props2} episodes={20}
      followedShows={followedShows2}
      updateEpisode={updateEpisode}
      updateComment={updateComment}/>
      )
      expect(wrapper.instance().setComment()).toEqual('Testing')
  })
})