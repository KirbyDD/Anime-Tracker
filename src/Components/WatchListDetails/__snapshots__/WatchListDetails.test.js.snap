// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WatchList Details should match snapshot 1`] = `
<section
  className="watchlist-details"
>
  <div
    className="watchlist-detail-info"
  >
    <h2>
      Life of TronKat
    </h2>
    <h3>
      This is the synopsis
    </h3>
    <p
      className="detail-text"
    >
      What Episode are you currently on?
    </p>
    <p
      className="detail-text"
    />
    <button
      onClick={[Function]}
    >
      Update
    </button>
    <p
      className="detail-text"
    >
      Reminder Section
    </p>
    <p
      className="detail-text"
    />
    <button
      onClick={[Function]}
    >
      Update
    </button>
    <br />
    <Link
      to="/watchlist"
    >
      <button
        className="remove-btn"
        onClick={[Function]}
      >
        Remove From WatchList
      </button>
    </Link>
  </div>
  <img
    className="side-image"
    src="https://i.imgur.com/vihivug.jpg"
  />
</section>
`;

exports[`WatchList Details should match snapshot 2`] = `
<section
  className="watchlist-details"
>
  <div
    className="watchlist-detail-info"
  >
    <h2>
      Life of TronKat
    </h2>
    <h3>
      This is the synopsis
    </h3>
    <p
      className="detail-text"
    >
      What Episode are you currently on?
    </p>
    <p
      className="detail-text"
    />
    <select
      onChange={[Function]}
      selected={1}
    >
      <option
        value={1}
      >
        1
      </option>
      <option
        value={2}
      >
        2
      </option>
      <option
        value={3}
      >
        3
      </option>
      <option
        value={4}
      >
        4
      </option>
      <option
        value={5}
      >
        5
      </option>
      <option
        value={6}
      >
        6
      </option>
      <option
        value={7}
      >
        7
      </option>
      <option
        value={8}
      >
        8
      </option>
      <option
        value={9}
      >
        9
      </option>
      <option
        value={10}
      >
        10
      </option>
      <option
        value={11}
      >
        11
      </option>
      <option
        value={12}
      >
        12
      </option>
      <option
        value={13}
      >
        13
      </option>
      <option
        value={14}
      >
        14
      </option>
      <option
        value={15}
      >
        15
      </option>
      <option
        value={16}
      >
        16
      </option>
      <option
        value={17}
      >
        17
      </option>
      <option
        value={18}
      >
        18
      </option>
      <option
        value={19}
      >
        19
      </option>
      <option
        value={20}
      >
        20
      </option>
    </select>
    <p
      className="detail-text"
    >
      Reminder Section
    </p>
    <p
      className="detail-text"
    />
    <button
      onClick={[Function]}
    >
      Update
    </button>
    <br />
    <Link
      to="/watchlist"
    >
      <button
        className="remove-btn"
        onClick={[Function]}
      >
        Remove From WatchList
      </button>
    </Link>
  </div>
  <img
    className="side-image"
    src="https://i.imgur.com/vihivug.jpg"
  />
</section>
`;

exports[`WatchList Details should match snapshot 3`] = `
<section
  className="watchlist-details"
>
  <div
    className="watchlist-detail-info"
  >
    <h2>
      Life of TronKat
    </h2>
    <h3>
      This is the synopsis
    </h3>
    <p
      className="detail-text"
    >
      What Episode are you currently on?
    </p>
    <p
      className="detail-text"
    />
    <button
      onClick={[Function]}
    >
      Update
    </button>
    <p
      className="detail-text"
    >
      Reminder Section
    </p>
    <p
      className="detail-text"
    />
    <article>
      <textarea
        placeholder="Type new reminder here"
      />
      <br />
      <button
        onClick={[Function]}
      >
        Sumbit
      </button>
    </article>
    <br />
    <Link
      to="/watchlist"
    >
      <button
        className="remove-btn"
        onClick={[Function]}
      >
        Remove From WatchList
      </button>
    </Link>
  </div>
  <img
    className="side-image"
    src="https://i.imgur.com/vihivug.jpg"
  />
</section>
`;
